src/tpm2-crypto.c

```diff
@@ -529,13 +529,26 @@ static UsefulBuf concat_data_sign(UsefulBuf data, UsefulBuf sign, UsefulBuf buf)
     }
 }
 
-UsefulBuf sign_with_aik(UsefulBuf data)
+UsefulBuf sign_with_aik(UsefulBuf data, UsefulBuf nonce)
 {
     UsefulBuf   ret = NULLUsefulBuf;
     UsefulBuf   signature = NULLUsefulBuf;
+    UsefulBuf   data_and_nonce = NULLUsefulBuf;
 
-    signature = get_aik_signature(data);
-    if (UsefulBuf_IsNULLOrEmpty(data)) {
+    if (!UsefulBuf_IsNULLOrEmpty(nonce)) {
+        data_and_nonce.len = data.len + nonce.len;
+        data_and_nonce.ptr = malloc(data_and_nonce.len);
+        memcpy(data_and_nonce.ptr, data.ptr, data.len);
+        memcpy((uint8_t*)data_and_nonce.ptr + data.len, nonce.ptr, nonce.len);
+    }
+
+    if (UsefulBuf_IsNULLOrEmpty(data_and_nonce)) {
+        signature = get_aik_signature(data);
+    } else {
+        signature = get_aik_signature(data_and_nonce);
+        free(data_and_nonce.ptr);
+    }
```

Comment:
```suggestion
    if (!UsefulBuf_IsNULLOrEmpty(nonce)) {
        data_and_nonce.len = data.len + nonce.len;
        data_and_nonce.ptr = malloc(data_and_nonce.len);
        memcpy(data_and_nonce.ptr, data.ptr, data.len);
        memcpy((uint8_t*)data_and_nonce.ptr + data.len, nonce.ptr, nonce.len);

        signature = get_aik_signature(data_and_nonce);
        free(data_and_nonce.ptr);
    } else {
        signature = get_aik_signature(data);
    }
```
`data_and_nonce` can also be limited to this block.